core/src/main/scala/kafka/common/InterBrokerSendThread.scala

```diff
@@ -84,9 +74,130 @@ abstract class InterBrokerSendThread(name: String,
     }
   }
 
-  def wakeup(): Unit = networkClient.wakeup()
+  private def sendRequests(now: Long): Unit = {
+    var pollTimeout = Long.MaxValue
+    if (unsentRequests.hasRequests) {
+      for (node <- unsentRequests.nodes.asScala) {
+        val requestIterator = unsentRequests.requestIterator(node)
+        while (requestIterator.hasNext) {
+          val request = requestIterator.next
+          if (networkClient.ready(node, now)) {
+            networkClient.send(request, now)
+            requestIterator.remove()
+          } else
+            pollTimeout = if (unsentRequests.hasRequests) AdminClient.DefaultReconnectBackoffMs else Long.MaxValue
+        }
+      }
+    }
+    networkClient.poll(pollTimeout, now)
+  }
+
+  private def checkDisconnects(now: Long): Unit = {
+    // any disconnects affecting requests that have already been transmitted will be handled
+    // by NetworkClient, so we just need to check whether connections for any of the unsent
+    // requests have been disconnected; if they have, then we complete the corresponding future
+    // and set the disconnect flag in the ClientResponse
+    val nodeIterator = unsentRequests.nodeIterator()
+    while (nodeIterator.hasNext) {
+      val node = nodeIterator.next
+      if (networkClient.connectionFailed(node)) {
+        // Remove entry before invoking request callback to avoid callbacks handling
+        // coordinator failures traversing the unsent list again.
+        val requests = unsentRequests.getRequests(node)
+        nodeIterator.remove()
+        for (request <- requests.asScala) {
+          val handler = request.callback
+          if (networkClient.authenticationException(node) == null)
+            error(s"Failed to send the following request due to authentication error: ${request.toString}")
+          handler.onComplete(new ClientResponse(request.makeHeader(request.requestBuilder().latestAllowedVersion()),
+            handler, request.destination, now /* createdTimeMs */ , now /* receivedTimeMs */ , true /* disconnected */ ,
+            null /* versionMismatch */ , null /* responseBody */))
+        }
+      }
+    }
+  }
 
+  private def failExpiredRequests(now: Long): Unit = {
+    // clear all expired unsent requests
+    val expiredRequests = unsentRequests.removeExpiredRequests(now, unsentExpiryMs)
+    for (request <- expiredRequests.asScala) {
+      debug(s"Failed to send the following request after $unsentExpiryMs ms: ${request.toString}")
+      val handler = request.callback
+      handler.onComplete(new ClientResponse(request.makeHeader(request.requestBuilder().latestAllowedVersion()),
```

Comment:
nit: maybe we could put this into a method since we have the same logic above.

Reply:
Done!